### PR TITLE
fix: Added error handler for status 429

### LINKED ISSUE #6659: Add error handler for status 429
Status 429 Too many requests has no error handler, and hence it returns HTML instead of JSON

https://flask-limiter.readthedocs.io/en/stable/#custom-rate-limit-exceeded-responses

### PR DESCRIPTION
**Pull request by Rits1272**
Fixes #6659 

#### Short description of what this resolves: Added Error handler for status 429 in the errors.py file

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by Rits1272** [comment]
Please review @codedsun

**Comment by Rits1272** [comment]
Kindly review again @codedsun

**Comment by kushthedude** [comment]
Also, Squash your commits.

**Comment by Rits1272** [comment]
Sure @kushthedude and thanks for reviewing @codedsun, @kushthedude  and @mrsaicharan1!

**Comment by Rits1272** [comment]
@kushthedude I have changed the PR title in accordance with semantic PR naming and also squashed the commits.

**Comment by Rits1272** [comment]
@codedsun can you please tell in which file should I register the error handler?

**Comment by Rits1272** [comment]
@codedsun can you please check if I have registered the handler in the app correctly?

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843&bid=15715326)

File: app/__init__.py [hunk 1]
```diff
@@ -287,5 +289,9 @@ def internal_server_error(error):
                          {'Content-Type': 'application/vnd.api+json'})
 
 
+@app.errorhandler(429)
+def request_limit_error(error):
+    raise RequestLimitError({'source': ''}, error.description).respond()
```
Comment (niranjan94) [inline app/__init__.py]: Codacy Issue found: [undefined name 'RequestLimitError'](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

**Comment by Rits1272** [comment]
Please @codedsun check. I think I have registered the error handler successfully

File: app/api/helpers/errors.py [hunk 2]
```diff
@@ -1,8 +1,9 @@
 import json
```
Comment (kushthedude) [inline app/api/helpers/errors.py]: Why?

File: app/api/helpers/errors.py [hunk 3]
```diff
@@ -78,3 +79,16 @@ class BadRequestError(ErrorResponse):
...
+    """
+    Default class for 429 Error
+    """
+    status = 429
+    title = 'Request Limit Exceeded'
+
+
+@error_blueprint.app_errorhandler(429)
+def handle_429(error):
+    raise RequestLimitError({'source': ''}, error.description).respond()
```
Comment (kushthedude) [inline app/api/helpers/errors.py]: Why this? Keep it consistent with other handlers

Comment (Rits1272) [reply app/api/helpers/errors.py]: Can you please tell me where other error handlers are implemented?

File: app/api/helpers/errors.py (hunk 2 above)
Comment (Rits1272) [reply app/api/helpers/errors.py]: Will remove it

File: app/api/helpers/errors.py (hunk 3 above)
Comment (iamareebjamal) [reply app/api/helpers/errors.py]: They are present in __init__.py

File: app/api/helpers/errors.py [hunk 4]
```diff
@@ -78,3 +79,16 @@ class BadRequestError(ErrorResponse):
...
     status = 400
     title = 'Bad Request'
+
+
+class RequestLimitError(ErrorResponse):
+    """
+    Default class for 429 Error
+    """
+    status = 429
+    title = 'Request Limit Exceeded'
```
Comment (iamareebjamal) [inline app/api/helpers/errors.py]: Not needed. Try to understand the issue

File: tests/all/integration/api/helpers/test_errors.py [hunk 5]
```diff
@@ -47,6 +47,11 @@ def test_errors(self):
             bad_request_error = BadRequestError({'source': ''}, 'Request cannot be served')
             self.assertEqual(bad_request_error.status, 400)
 
+            # Request limit Error
+            request_limit_error = RequestLimitError({'source': ''},
+                                                    'Request limit exceeded')
+            self.assertEqual(request_limit_error.status, 429)
```
Comment (iamareebjamal) [inline tests/all/integration/api/helpers/test_errors.py]: Not needed

File: app/api/helpers/errors.py (hunk 3 above)
Comment (Rits1272) [reply app/api/helpers/errors.py]: Ok @iamareebjamal. Working on it. Thanks for reviewing!

File: app/api/helpers/errors.py (hunk 4 above)
Comment (Rits1272) [reply app/api/helpers/errors.py]: Sure

File: tests/all/integration/api/helpers/test_errors.py (hunk 5 above)
Comment (Rits1272) [reply tests/all/integration/api/helpers/test_errors.py]: Will remove it

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843&bid=15715326)

File: app/api/helpers/errors.py [hunk 6]
```diff
@@ -24,7 +22,7 @@ def __init__(self, source, detail, title=None, status=None):
         if title is not None:
             self.title = title
         if status is not None:
-            self.status = status
+            self.status = staRequestLimitErrortus
```
Comment (niranjan94) [inline app/api/helpers/errors.py]: Codacy Issue found: [undefined name 'staRequestLimitErrortus'](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843&bid=15715326)

File: app/api/helpers/errors.py [hunk 7]
```diff
@@ -77,4 +76,4 @@ class BadRequestError(ErrorResponse):
     Default class for 400 Error
     """
     status = 400
-    title = 'Bad Request'
+    title = 'Bad Request'
```
Comment (niranjan94) [inline app/api/helpers/errors.py]: Codacy Issue found: [Final newline missing](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

File: app/__init__.py [hunk 8]
```diff
@@ -287,5 +287,13 @@ def internal_server_error(error):
...
+def ratelimit_handler(err):
+    return make_response(
+            jsonify(error="ratelimit exceeded %s" % err.description)
+            ,429
+    )
+
+
 if __name__ == '__main__':
-    current_app.run()
+    current_app.run()
```
Comment (niranjan94) [inline app/__init__.py]: Codacy Issue found: [Final newline missing](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

File: app/__init__.py [hunk 9]
```diff
@@ -287,5 +287,13 @@ def internal_server_error(error):
                          {'Content-Type': 'application/vnd.api+json'})
 
 
+@app.errorhandler(429)
+def ratelimit_handler(err):
+    return make_response(
+            jsonify(error="ratelimit exceeded %s" % err.description)
```
Comment (iamareebjamal) [inline app/__init__.py]: Follow conventions of other errors

File: app/api/helpers/errors.py [hunk 10]
```diff
@@ -15,7 +15,6 @@ class ErrorResponse:
 
     def __init__(self, source, detail, title=None, status=None):
         """Initialize a jsonapi ErrorResponse Object
-
```
Comment (iamareebjamal) [inline app/api/helpers/errors.py]: revert

Comment (Rits1272) [reply app/api/helpers/errors.py]: Okay

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843&bid=15715326)

File: app/__init__.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
 from envparse import env
 
 import sys
-from flask import Flask, json, make_response
+from flask import Flask, json, make_response, jsonify
```
Comment (niranjan94) [inline app/__init__.py]: Codacy Issue found: ['flask.jsonify' imported but unused](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

File: app/api/helpers/errors.py [hunk 12]
```diff
@@ -15,7 +15,7 @@ class ErrorResponse:
...
+
```
Comment (niranjan94) [inline app/api/helpers/errors.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4703843)

File: app/__init__.py [hunk 13]
```diff
@@ -290,11 +290,10 @@ def internal_server_error(error):
...
 @app.errorhandler(429)
-def ratelimit_handler(err):
-    return make_response(
-            jsonify(error="ratelimit exceeded %s" % err.description)
-            , 429
-    )
+def ratelimit_handler(error):
+    exc = JsonApiException({'pointer': ''}, str(error))
+    return make_response(json.dumps(jsonapi_errors([exc.to_dict()])), exc.status,
+                        {'Content-Type': 'application/vnd.api+json'})
```
Comment (iamareebjamal) [inline app/__init__.py]: This is not what we want. Did you even test what it returns?

Comment (Rits1272) [reply app/__init__.py]: Ok, Working on it @iamareebjamal.

**Comment by Rits1272** [comment]
Please review @iamareebjamal

**Comment by iamareebjamal** [comment]
No need for jsonapi errors, source or details

**Comment by Rits1272** [comment]
Made the changes @iamareebjamal

File: app/instance.py [hunk 14]
```diff
@@ -261,5 +261,14 @@ def internal_server_error(error):
                          {'Content-Type': 'application/vnd.api+json'})
 
 
+@app.errorhandler(429)
+def ratelimit_handler(error):
+    dict_ = {'status': 429,
+             'title': 'Request Limit Exceeded',
+             }
+    return make_response(json.dumps(dict_), 429,
+                         {'Content-Type': 'application/vnd.api+json'})
```
Comment (iamareebjamal) [inline app/instance.py]: Properly format and indent. Remove dict_ variable

Comment (Rits1272) [reply app/instance.py]: Okay

**Comment by Rits1272** [comment]
I made the suggested changes @iamareebjamal

File: app/instance.py [hunk 15]
```diff
@@ -261,5 +261,16 @@ def internal_server_error(error):
                          {'Content-Type': 'application/vnd.api+json'})
 
 
+@app.errorhandler(429)
+def ratelimit_handler(error):
+    return make_response(json.dumps({
+            'status': 429,
+            'title': 'Request Limit Exceeded'
+        }),
+        429, {
```
Comment (iamareebjamal) [inline app/instance.py]: { should start from next line

**Comment by Rits1272** [comment]
Made the changes @iamareebjamal

**Comment by Rits1272** [comment]
Thanks a lot, @mrsaicharan1  and @iamareebjamal for having patience with me. I will try my best in improving my PR from next time.

**Pull request merged**